src/thermo/PengRobinson.cpp

```diff
@@ -0,0 +1,817 @@
+            molarVolLast = m_Vroot[0];
+        } else if (phaseRequested == FLUID_GAS || phaseRequested == FLUID_SUPERCRIT) {
+            molarVolLast = m_Vroot[2];
+        } else {
+            if (volGuess > m_Vroot[1]) {
+                molarVolLast = m_Vroot[2];
+            } else {
+                molarVolLast = m_Vroot[0];
+            }
+        }
+    } else if (m_NSolns == 1) {
+        if (phaseRequested == FLUID_GAS || phaseRequested == FLUID_SUPERCRIT || phaseRequested == FLUID_UNDEFINED) {
+            molarVolLast = m_Vroot[0];
+        } else {
+            return -2.0;
+        }
+    } else if (m_NSolns == -1) {
+        if (phaseRequested >= FLUID_LIQUID_0 || phaseRequested == FLUID_UNDEFINED || phaseRequested == FLUID_SUPERCRIT) {
+            molarVolLast = m_Vroot[0];
+        } else if (T > tcrit) {
+            molarVolLast = m_Vroot[0];
+        } else {
+            return -2.0;
+        }
+    } else {
+        molarVolLast = m_Vroot[0];
+        return -1.0;
+    }
+    return mmw / molarVolLast;
+}
+
+double PengRobinson::densSpinodalLiquid() const
+{
+    double Vroot[3];
+    double T = temperature();
+    int nsol = solveCubic(T, pressure(), m_a, m_b, m_aAlpha_mix, Vroot);
+    if (nsol != 3) {
+        return critDensity();
+    }
+
+    auto resid = [this, T](double v) {
+        double pp;
+        return dpdVCalc(T, v, pp);
+    };
+
+    boost::uintmax_t maxiter = 100;
+    std::pair<double, double> vv = bmt::toms748_solve(
+        resid, Vroot[0], Vroot[1], bmt::eps_tolerance<double>(48), maxiter);
+
+    double mmw = meanMolecularWeight();
+    return mmw / (0.5 * (vv.first + vv.second));
+}
+
+double PengRobinson::densSpinodalGas() const
+{
+    double Vroot[3];
+    double T = temperature();
+    int nsol = solveCubic(T, pressure(), m_a, m_b, m_aAlpha_mix, Vroot);
+    if (nsol != 3) {
+        return critDensity();
+    }
+
+    auto resid = [this, T](double v) {
+        double pp;
+        return dpdVCalc(T, v, pp);
+    };
+
+    boost::uintmax_t maxiter = 100;
+    std::pair<double, double> vv = bmt::toms748_solve(
+        resid, Vroot[1], Vroot[2], bmt::eps_tolerance<double>(48), maxiter);
+
+    double mmw = meanMolecularWeight();
+    return mmw / (0.5 * (vv.first + vv.second));
+}
+
+double PengRobinson::dpdVCalc(double T, double molarVol, double& presCalc) const
+{
+    double den = molarVol * molarVol + 2 * molarVol * m_b - m_b * m_b;
+    presCalc = GasConstant * T / (molarVol - m_b) - m_aAlpha_mix / den;
+
+    double vpb = molarVol + m_b;
+    double vmb = molarVol - m_b;
+    double dpdv = -GasConstant * T / (vmb * vmb) + 2 * m_aAlpha_mix * vpb / (den*den);
+    return dpdv;
+}
+
+void PengRobinson::calculatePressureDerivatives() const
+{
+    double T = temperature();
+    double mv = molarVolume();
+    double pres;
+
+    m_dpdV = dpdVCalc(T, mv, pres);
+    double vmb = mv - m_b;
+    double den = mv * mv + 2 * mv * m_b - m_b * m_b;
```

Comment:
Might be clearer to call this `denom` for "denominator"... `den` is easy to confuse with "density."

Reply:
Corrected.